File: archive/cdk-java/Pr2246/lib/src/main/java/app/Route53Stack.java
package app;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.route53.*;
import software.constructs.Construct;

import java.util.Optional;

/**
 * Route53StackProps holds configuration for the Route53Stack.
 */
class Route53StackProps {
    private final String environmentSuffix;
    private final StackProps stackProps;

    private Route53StackProps(String environmentSuffix, StackProps stackProps) {
        this.environmentSuffix = environmentSuffix;
        this.stackProps = stackProps != null ? stackProps : StackProps.builder().build();
    }

    public String getEnvironmentSuffix() {
        return environmentSuffix;
    }

    public StackProps getStackProps() {
        return stackProps;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private String environmentSuffix;
        private StackProps stackProps;

        public Builder environmentSuffix(String environmentSuffix) {
            this.environmentSuffix = environmentSuffix;
            return this;
        }

        public Builder stackProps(StackProps stackProps) {
            this.stackProps = stackProps;
            return this;
        }

        public Route53StackProps build() {
            return new Route53StackProps(environmentSuffix, stackProps);
        }
    }
}

/**
 * Route53Stack creates DNS failover configuration for multi-region deployment.
 * This provides automatic DNS failover between regions based on health checks.
 */
public class Route53Stack extends Stack {
    private final String environmentSuffix;

    public Route53Stack(final Construct scope, final String id, final Route53StackProps props) {
        super(scope, id, props != null ? props.getStackProps() : null);

        this.environmentSuffix = Optional.ofNullable(props)
                .map(Route53StackProps::getEnvironmentSuffix)
                .orElse("dev");

        // Create hosted zone
        HostedZone hostedZone = HostedZone.Builder.create(this, "WebAppHostedZone" + environmentSuffix)
                .zoneName("myapp-" + environmentSuffix + ".example.org")
                .comment("Hosted zone for multi-region web application")
                .build();

        // Note: In a real implementation, you would reference the ALBs from other stacks
        // For this example, we'll create placeholder health checks and records
        
        // Health check for primary region (us-east-1)
        CfnHealthCheck primaryHealthCheck = CfnHealthCheck.Builder.create(this, "PrimaryHealthCheck" + environmentSuffix)
                .healthCheckConfig(CfnHealthCheck.HealthCheckConfigProperty.builder()
                        .type("HTTP")
                        .resourcePath("/")
                        .fullyQualifiedDomainName("primary-alb-" + environmentSuffix + ".us-east-1.elb.amazonaws.com")
                        .port(80)
                        .requestInterval(30)
                        .failureThreshold(3)
                        .build())
                .build();

        // Health check for secondary region (us-west-2)
        CfnHealthCheck secondaryHealthCheck = CfnHealthCheck.Builder.create(this, "SecondaryHealthCheck" + environmentSuffix)
                .healthCheckConfig(CfnHealthCheck.HealthCheckConfigProperty.builder()
                        .type("HTTP")
                        .resourcePath("/")
                        .fullyQualifiedDomainName("secondary-alb-" + environmentSuffix + ".us-west-2.elb.amazonaws.com")
                        .port(80)
                        .requestInterval(30)
                        .failureThreshold(3)
                        .build())
                .build();

        // Primary record (us-east-1)
        // Using latency-based routing instead of geo-routing
        ARecord.Builder.create(this, "PrimaryRecord" + environmentSuffix)
                .zone(hostedZone)
                .recordName("www")
                .target(RecordTarget.fromIpAddresses("192.0.2.1")) // Placeholder IP
                .setIdentifier("primary")
                .region("us-east-1") // Use latency-based routing with region
                .build();

        // Secondary record (us-west-2)
        // Using latency-based routing instead of geo-routing
        ARecord.Builder.create(this, "SecondaryRecord" + environmentSuffix)
                .zone(hostedZone)
                .recordName("www")
                .target(RecordTarget.fromIpAddresses("192.0.2.2")) // Placeholder IP
                .setIdentifier("secondary")
                .region("us-west-2") // Use latency-based routing with region
                .build();
    }
}

File: archive/cdk-py/Pr6928/lib/route53_stack.py
"""route53_stack.py
Route 53 health checks and weighted routing for failover.
"""

from typing import Optional
import aws_cdk as cdk
from aws_cdk import aws_route53 as route53
from constructs import Construct


class Route53StackProps:
    """Properties for Route53 stack."""

    def __init__(
        self,
        environment_suffix: str,
        primary_function_url: str,
        secondary_function_url: str
    ):
        self.environment_suffix = environment_suffix
        self.primary_function_url = primary_function_url
        self.secondary_function_url = secondary_function_url


class Route53Stack(Construct):
    """Creates Route 53 health checks and routing policies."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        props: Route53StackProps
    ):
        super().__init__(scope, construct_id)

        # Create hosted zone (assuming domain is provided)
        self.hosted_zone = route53.HostedZone(
            self,
            f'HostedZone{props.environment_suffix}',
            zone_name=f'dr-example-{props.environment_suffix}.com',
            comment='Hosted zone for disaster recovery solution'
        )

        # Create health check for primary region
        primary_health_check = route53.CfnHealthCheck(
            self,
            f'PrimaryHealthCheck{props.environment_suffix}',
            health_check_config=route53.CfnHealthCheck.HealthCheckConfigProperty(
                type='HTTPS',
                resource_path='/health',
                fully_qualified_domain_name=props.primary_function_url,
                port=443,
                request_interval=30,
                failure_threshold=3,
                measure_latency=True
            ),
            health_check_tags=[
                route53.CfnHealthCheck.HealthCheckTagProperty(
                    key='Name',
                    value=f'dr-primary-health-check-{props.environment_suffix}'
                ),
                route53.CfnHealthCheck.HealthCheckTagProperty(
                    key='DR-Role',
                    value='Primary-HealthCheck'
                )
            ]
        )

        # Create health check for secondary region
        secondary_health_check = route53.CfnHealthCheck(
            self,
            f'SecondaryHealthCheck{props.environment_suffix}',
            health_check_config=route53.CfnHealthCheck.HealthCheckConfigProperty(
                type='HTTPS',
                resource_path='/health',
                fully_qualified_domain_name=props.secondary_function_url,
                port=443,
                request_interval=30,
                failure_threshold=3,
                measure_latency=True
            ),
            health_check_tags=[
                route53.CfnHealthCheck.HealthCheckTagProperty(
                    key='Name',
                    value=f'dr-secondary-health-check-{props.environment_suffix}'
                ),
                route53.CfnHealthCheck.HealthCheckTagProperty(
                    key='DR-Role',
                    value='Secondary-HealthCheck'
                )
            ]
        )

        # Create A record with weighted routing for primary
        route53.ARecord(
            self,
            f'PrimaryARecord{props.environment_suffix}',
            zone=self.hosted_zone,
            record_name='api',
            target=route53.RecordTarget.from_ip_addresses(props.primary_function_url),
            weight=100,
            set_identifier='primary'
        )

        # Create A record with weighted routing for secondary
        route53.ARecord(
            self,
            f'SecondaryARecord{props.environment_suffix}',
            zone=self.hosted_zone,
            record_name='api',
            target=route53.RecordTarget.from_ip_addresses(props.secondary_function_url),
            weight=0,
            set_identifier='secondary'
        )

        # Outputs
        cdk.CfnOutput(
            self,
            'HostedZoneId',
            value=self.hosted_zone.hosted_zone_id,
            description='Route 53 hosted zone ID'
        )
        cdk.CfnOutput(
            self,
            'HostedZoneName',
            value=self.hosted_zone.zone_name,
            description='Route 53 hosted zone name'
        )